ENUM with DEFAULT NULL

- Create an enum
- Create a nullable field that uses this enum
- Create a controller

Compilation fails with

```
build/Generated/Types.hs:156:132: error: parse error on input `='
    |
156 |     fromField field (Just value) = returnError ConversionFailed field "Unexpected value for enum value"    fromField field Nothing = returnError UnexpectedNull field "Unexpected null for enum value"
 ```

--- a/IHP/ErrorController.hs
+++ b/IHP/ErrorController.hs
@@ -71,7 +71,7 @@ postgresHandler exception controller additionalInfo = do
                         <h2>Possible Solutions</h2>
                         <div style="margin-bottom: 2rem; font-weight: 400;">
                             Have you clicked on
-                            <form method="POST" action="http://localhost:8001/ihp/PushToDb" target="_blank" style="display: inline">
+                            <form method="POST" action="http://localhost:8001/PushToDb" target="_blank" style="display: inline">
                                 <button type="submit">Push to DB</button>
                             </form>
                             after updating the Schema?

--- a/IHP/IDE/SchemaDesigner/Compiler.hs
+++ b/IHP/IDE/SchemaDesigner/Compiler.hs
@@ -3,12 +3,13 @@ Module: IHP.IDE.SchemaDesigner.Compiler
 Description: Compiles AST of SQL to DDL
 Copyright: (c) digitally induced GmbH, 2020
 -}
-module IHP.IDE.SchemaDesigner.Compiler (compileSql, writeSchema, compileIdentifier) where
+module IHP.IDE.SchemaDesigner.Compiler (compileSql, writeSchema, compileIdentifier, compileExpression) where
 
 import IHP.Prelude
 import IHP.IDE.SchemaDesigner.Types
 import Data.Maybe (fromJust)
 import qualified Data.Text.IO as Text
+import qualified Data.Text as Text
 
 writeSchema :: [Statement] -> IO ()
 writeSchema !statements = do
@@ -22,7 +23,7 @@ compileSql statements = statements
 
 compileStatement :: Statement -> Text
 compileStatement CreateTable { name, columns } = "CREATE TABLE " <> compileIdentifier name <> " (\n" <> intercalate ",\n" (map compileColumn columns) <> "\n);"
-compileStatement CreateEnumType { name, values } = "CREATE TYPE " <> compileIdentifier name <> " AS ENUM (" <> intercalate ", " values <> ");"
+compileStatement CreateEnumType { name, values } = "CREATE TYPE " <> compileIdentifier name <> " AS ENUM (" <> intercalate ", " (values |> map TextExpression |> map compileExpression) <> ");"
 compileStatement CreateExtension { name, ifNotExists } = "CREATE EXTENSION " <> (if ifNotExists then "IF NOT EXISTS " else "") <> "\"" <> compileIdentifier name <> "\";"
 compileStatement AddConstraint { tableName, constraintName, constraint } = "ALTER TABLE " <> compileIdentifier tableName <> " ADD CONSTRAINT " <> compileIdentifier constraintName <> " " <> compileConstraint constraint <> ";"
 compileStatement Comment { content } = "-- " <> content
@@ -49,255 +50,264 @@ compileColumn Column { name, columnType, primaryKey, defaultValue, notNull, isUn
         , if isUnique then Just "UNIQUE" else Nothing
         ])
 
-compileDefaultValue :: Text -> Text
-compileDefaultValue value = "DEFAULT " <> value
+compileDefaultValue :: Expression -> Text
+compileDefaultValue value = "DEFAULT " <> compileExpression value
+
+compileExpression :: Expression -> Text
+compileExpression (TextExpression value) = "'" <> value <> "'"
+compileExpression (VarExpression name) = name
+compileExpression (CallExpression func args) = func <> "(" <> intercalate ", " (map compileExpression args) <> ")"
 
 compareStatement (CreateTable {}) _ = LT
 compareStatement (AddConstraint {}) _ = GT
 compareStatement _ _ = EQ
 
 compileIdentifier :: _ -> Text
-compileIdentifier keyword = case (IHP.Prelude.toUpper keyword) of
-    "ABORT" -> tshow keyword
-    "ABSOLUTE" -> tshow keyword
-    "ACCESS" -> tshow keyword
-    "ACTION" -> tshow keyword
-    "ADD" -> tshow keyword
-    "ADMIN" -> tshow keyword
-    "AFTER" -> tshow keyword
-    "AGGREGATE" -> tshow keyword
-    "ALSO" -> tshow keyword
-    "ALTER" -> tshow keyword
-    "ASSERTION" -> tshow keyword
-    "ASSIGNMENT" -> tshow keyword
-    "AT" -> tshow keyword
-    "BACKWARD" -> tshow keyword
-    "BEFORE" -> tshow keyword
-    "BEGIN" -> tshow keyword
-    "BY" -> tshow keyword
-    "CACHE" -> tshow keyword
-    "CALLED" -> tshow keyword
-    "CASCADE" -> tshow keyword
-    "CHAIN" -> tshow keyword
-    "CHARACTERISTICS" -> tshow keyword
-    "CHECKPOINT" -> tshow keyword
-    "CLASS" -> tshow keyword
-    "CLOSE" -> tshow keyword
-    "CLUSTER" -> tshow keyword
-    "COMMENT" -> tshow keyword
-    "COMMIT" -> tshow keyword
-    "COMMITTED" -> tshow keyword
-    "CONNECTION" -> tshow keyword
-    "CONSTRAINTS" -> tshow keyword
-    "CONVERSION" -> tshow keyword
-    "COPY" -> tshow keyword
-    "CREATEDB" -> tshow keyword
-    "CREATEROLE" -> tshow keyword
-    "CREATEUSER" -> tshow keyword
-    "CSV" -> tshow keyword
-    "CURSOR" -> tshow keyword
-    "CYCLE" -> tshow keyword
-    "DATABASE" -> tshow keyword
-    "DAY" -> tshow keyword
-    "DEALLOCATE" -> tshow keyword
-    "DECLARE" -> tshow keyword
-    "DEFAULTS" -> tshow keyword
-    "DEFERRED" -> tshow keyword
-    "DEFINER" -> tshow keyword
-    "DELETE" -> tshow keyword
-    "DELIMITER" -> tshow keyword
-    "DELIMITERS" -> tshow keyword
-    "DISABLE" -> tshow keyword
-    "DOMAIN" -> tshow keyword
-    "DOUBLE" -> tshow keyword
-    "DROP" -> tshow keyword
-    "EACH" -> tshow keyword
-    "ENABLE" -> tshow keyword
-    "ENCODING" -> tshow keyword
-    "ENCRYPTED" -> tshow keyword
-    "ESCAPE" -> tshow keyword
-    "EXCLUDING" -> tshow keyword
-    "EXCLUSIVE" -> tshow keyword
-    "EXECUTE" -> tshow keyword
-    "EXPLAIN" -> tshow keyword
-    "EXTERNAL" -> tshow keyword
-    "FETCH" -> tshow keyword
-    "FIRST" -> tshow keyword
-    "FORCE" -> tshow keyword
-    "FORWARD" -> tshow keyword
-    "FUNCTION" -> tshow keyword
-    "GLOBAL" -> tshow keyword
-    "GRANTED" -> tshow keyword
-    "HANDLER" -> tshow keyword
-    "HEADER" -> tshow keyword
-    "HOLD" -> tshow keyword
-    "HOUR" -> tshow keyword
-    "IMMEDIATE" -> tshow keyword
-    "IMMUTABLE" -> tshow keyword
-    "IMPLICIT" -> tshow keyword
-    "INCLUDING" -> tshow keyword
-    "INCREMENT" -> tshow keyword
-    "INDEX" -> tshow keyword
-    "INHERIT" -> tshow keyword
-    "INHERITS" -> tshow keyword
-    "INPUT" -> tshow keyword
-    "INSENSITIVE" -> tshow keyword
-    "INSERT" -> tshow keyword
-    "INSTEAD" -> tshow keyword
-    "INVOKER" -> tshow keyword
-    "ISOLATION" -> tshow keyword
-    "KEY" -> tshow keyword
-    "LANCOMPILER" -> tshow keyword
-    "LANGUAGE" -> tshow keyword
-    "LARGE" -> tshow keyword
-    "LAST" -> tshow keyword
-    "LEVEL" -> tshow keyword
-    "LISTEN" -> tshow keyword
-    "LOAD" -> tshow keyword
-    "LOCAL" -> tshow keyword
-    "LOCATION" -> tshow keyword
-    "LOCK" -> tshow keyword
-    "LOGIN" -> tshow keyword
-    "MATCH" -> tshow keyword
-    "MAXVALUE" -> tshow keyword
-    "MINUTE" -> tshow keyword
-    "MINVALUE" -> tshow keyword
-    "MODE" -> tshow keyword
-    "MONTH" -> tshow keyword
-    "MOVE" -> tshow keyword
-    "NAMES" -> tshow keyword
-    "NEXT" -> tshow keyword
-    "NO" -> tshow keyword
-    "NOCREATEDB" -> tshow keyword
-    "NOCREATEROLE" -> tshow keyword
-    "NOCREATEUSER" -> tshow keyword
-    "NOINHERIT" -> tshow keyword
-    "NOLOGIN" -> tshow keyword
-    "NOSUPERUSER" -> tshow keyword
-    "NOTHING" -> tshow keyword
-    "NOTIFY" -> tshow keyword
-    "NOWAIT" -> tshow keyword
-    "OBJECT" -> tshow keyword
-    "OF" -> tshow keyword
-    "OIDS" -> tshow keyword
-    "OPERATOR" -> tshow keyword
-    "OPTION" -> tshow keyword
-    "OWNER" -> tshow keyword
-    "PARTIAL" -> tshow keyword
-    "PASSWORD" -> tshow keyword
-    "PREPARE" -> tshow keyword
-    "PREPARED" -> tshow keyword
-    "PRESERVE" -> tshow keyword
-    "PRIOR" -> tshow keyword
-    "PRIVILEGES" -> tshow keyword
-    "PROCEDURAL" -> tshow keyword
-    "PROCEDURE" -> tshow keyword
-    "QUOTE" -> tshow keyword
-    "READ" -> tshow keyword
-    "RECHECK" -> tshow keyword
-    "REINDEX" -> tshow keyword
-    "RELATIVE" -> tshow keyword
-    "RELEASE" -> tshow keyword
-    "RENAME" -> tshow keyword
-    "REPEATABLE" -> tshow keyword
-    "REPLACE" -> tshow keyword
-    "RESET" -> tshow keyword
-    "RESTART" -> tshow keyword
-    "RESTRICT" -> tshow keyword
-    "RETURNS" -> tshow keyword
-    "REVOKE" -> tshow keyword
-    "ROLE" -> tshow keyword
-    "ROLLBACK" -> tshow keyword
-    "ROWS" -> tshow keyword
-    "RULE" -> tshow keyword
-    "SAVEPOINT" -> tshow keyword
-    "SCHEMA" -> tshow keyword
-    "SCROLL" -> tshow keyword
-    "SECOND" -> tshow keyword
-    "SECURITY" -> tshow keyword
-    "SEQUENCE" -> tshow keyword
-    "SERIALIZABLE" -> tshow keyword
-    "SESSION" -> tshow keyword
-    "SET" -> tshow keyword
-    "SHARE" -> tshow keyword
-    "tshow" -> tshow keyword
-    "SIMPLE" -> tshow keyword
-    "STABLE" -> tshow keyword
-    "START" -> tshow keyword
-    "STATEMENT" -> tshow keyword
-    "STATISTICS" -> tshow keyword
-    "STDIN" -> tshow keyword
-    "STDOUT" -> tshow keyword
-    "STORAGE" -> tshow keyword
-    "STRICT" -> tshow keyword
-    "SUPERUSER" -> tshow keyword
-    "SYSID" -> tshow keyword
-    "SYSTEM" -> tshow keyword
-    "TABLESPACE" -> tshow keyword
-    "TEMP" -> tshow keyword
-    "TEMPLATE" -> tshow keyword
-    "TEMPORARY" -> tshow keyword
-    "TOAST" -> tshow keyword
-    "TRANSACTION" -> tshow keyword
-    "TRIGGER" -> tshow keyword
-    "TRUNCATE" -> tshow keyword
-    "TRUSTED" -> tshow keyword
-    "TYPE" -> tshow keyword
-    "UNCOMMITTED" -> tshow keyword
-    "UNENCRYPTED" -> tshow keyword
-    "UNKNOWN" -> tshow keyword
-    "UNLISTEN" -> tshow keyword
-    "UNTIL" -> tshow keyword
-    "UPDATE" -> tshow keyword
-    "VACUUM" -> tshow keyword
-    "VALID" -> tshow keyword
-    "VALIDATOR" -> tshow keyword
-    "VALUES" -> tshow keyword
-    "VARYING" -> tshow keyword
-    "VIEW" -> tshow keyword
-    "VOLATILE" -> tshow keyword
-    "WITH" -> tshow keyword
-    "WITHOUT" -> tshow keyword
-    "WORK" -> tshow keyword
-    "WRITE" -> tshow keyword
-    "YEAR" -> tshow keyword
-    "ZONE" -> tshow keyword
-    "BIGINT" -> tshow keyword
-    "BIT" -> tshow keyword
-    "BOOLEAN" -> tshow keyword
-    "CHAR" -> tshow keyword
-    "CHARACTER" -> tshow keyword
-    "COALESCE" -> tshow keyword
-    "CONVERT" -> tshow keyword
-    "DEC" -> tshow keyword
-    "DECIMAL" -> tshow keyword
-    "EXISTS" -> tshow keyword
-    "EXTRACT" -> tshow keyword
-    "FLOAT" -> tshow keyword
-    "GREATEST" -> tshow keyword
-    "INOUT" -> tshow keyword
-    "INT" -> tshow keyword
-    "INTEGER" -> tshow keyword
-    "INTERVAL" -> tshow keyword
-    "LEAST" -> tshow keyword
-    "NATIONAL" -> tshow keyword
-    "NCHAR" -> tshow keyword
-    "NONE" -> tshow keyword
-    "NULLIF" -> tshow keyword
-    "NUMERIC" -> tshow keyword
-    "OUT" -> tshow keyword
-    "OVERLAY" -> tshow keyword
-    "POSITION" -> tshow keyword
-    "PRECISION" -> tshow keyword
-    "REAL" -> tshow keyword
-    "ROW" -> tshow keyword
-    "SETOF" -> tshow keyword
-    "SMALLINT" -> tshow keyword
-    "SUBSTRING" -> tshow keyword
-    "TIME" -> tshow keyword
-    "TIMESTAMP" -> tshow keyword
-    "TREAT" -> tshow keyword
-    "TRIM" -> tshow keyword
-    "VARCHAR" -> tshow keyword
-    _ -> cs keyword
+compileIdentifier identifier = if identifierNeedsQuoting then tshow identifier else identifier
+    where
+        identifierNeedsQuoting = isKeyword || containsSpace
+        isKeyword = IHP.Prelude.toUpper identifier `elem` keywords
+        containsSpace = Text.any (' ' ==) identifier
     
+        keywords = [ "ABORT"
+            , "ABSOLUTE"
+            , "ACCESS"
+            , "ACTION"
+            , "ADD"
+            , "ADMIN"
+            , "AFTER"
+            , "AGGREGATE"
+            , "ALSO"
+            , "ALTER"
+            , "ASSERTION"
+            , "ASSIGNMENT"
+            , "AT"
+            , "BACKWARD"
+            , "BEFORE"
+            , "BEGIN"
+            , "BY"
+            , "CACHE"
+            , "CALLED"
+            , "CASCADE"
+            , "CHAIN"
+            , "CHARACTERISTICS"
+            , "CHECKPOINT"
+            , "CLASS"
+            , "CLOSE"
+            , "CLUSTER"
+            , "COMMENT"
+            , "COMMIT"
+            , "COMMITTED"
+            , "CONNECTION"
+            , "CONSTRAINTS"
+            , "CONVERSION"
+            , "COPY"
+            , "CREATEDB"
+            , "CREATEROLE"
+            , "CREATEUSER"
+            , "CSV"
+            , "CURSOR"
+            , "CYCLE"
+            , "DATABASE"
+            , "DAY"
+            , "DEALLOCATE"
+            , "DECLARE"
+            , "DEFAULTS"
+            , "DEFERRED"
+            , "DEFINER"
+            , "DELETE"
+            , "DELIMITER"
+            , "DELIMITERS"
+            , "DISABLE"
+            , "DOMAIN"
+            , "DOUBLE"
+            , "DROP"
+            , "EACH"
+            , "ENABLE"
+            , "ENCODING"
+            , "ENCRYPTED"
+            , "ESCAPE"
+            , "EXCLUDING"
+            , "EXCLUSIVE"
+            , "EXECUTE"
+            , "EXPLAIN"
+            , "EXTERNAL"
+            , "FETCH"
+            , "FIRST"
+            , "FORCE"
+            , "FORWARD"
+            , "FUNCTION"
+            , "GLOBAL"
+            , "GRANTED"
+            , "HANDLER"
+            , "HEADER"
+            , "HOLD"
+            , "HOUR"
+            , "IMMEDIATE"
+            , "IMMUTABLE"
+            , "IMPLICIT"
+            , "INCLUDING"
+            , "INCREMENT"
+            , "INDEX"
+            , "INHERIT"
+            , "INHERITS"
+            , "INPUT"
+            , "INSENSITIVE"
+            , "INSERT"
+            , "INSTEAD"
+            , "INVOKER"
+            , "ISOLATION"
+            , "KEY"
+            , "LANCOMPILER"
+            , "LANGUAGE"
+            , "LARGE"
+            , "LAST"
+            , "LEVEL"
+            , "LISTEN"
+            , "LOAD"
+            , "LOCAL"
+            , "LOCATION"
+            , "LOCK"
+            , "LOGIN"
+            , "MATCH"
+            , "MAXVALUE"
+            , "MINUTE"
+            , "MINVALUE"
+            , "MODE"
+            , "MONTH"
+            , "MOVE"
+            , "NAMES"
+            , "NEXT"
+            , "NO"
+            , "NOCREATEDB"
+            , "NOCREATEROLE"
+            , "NOCREATEUSER"
+            , "NOINHERIT"
+            , "NOLOGIN"
+            , "NOSUPERUSER"
+            , "NOTHING"
+            , "NOTIFY"
+            , "NOWAIT"
+            , "OBJECT"
+            , "OF"
+            , "OIDS"
+            , "OPERATOR"
+            , "OPTION"
+            , "OWNER"
+            , "PARTIAL"
+            , "PASSWORD"
+            , "PREPARE"
+            , "PREPARED"
+            , "PRESERVE"
+            , "PRIOR"
+            , "PRIVILEGES"
+            , "PROCEDURAL"
+            , "PROCEDURE"
+            , "QUOTE"
+            , "READ"
+            , "RECHECK"
+            , "REINDEX"
+            , "RELATIVE"
+            , "RELEASE"
+            , "RENAME"
+            , "REPEATABLE"
+            , "REPLACE"
+            , "RESET"
+            , "RESTART"
+            , "RESTRICT"
+            , "RETURNS"
+            , "REVOKE"
+            , "ROLE"
+            , "ROLLBACK"
+            , "ROWS"
+            , "RULE"
+            , "SAVEPOINT"
+            , "SCHEMA"
+            , "SCROLL"
+            , "SECOND"
+            , "SECURITY"
+            , "SEQUENCE"
+            , "SERIALIZABLE"
+            , "SESSION"
+            , "SET"
+            , "SHARE"
+            , "tshow"
+            , "SIMPLE"
+            , "STABLE"
+            , "START"
+            , "STATEMENT"
+            , "STATISTICS"
+            , "STDIN"
+            , "STDOUT"
+            , "STORAGE"
+            , "STRICT"
+            , "SUPERUSER"
+            , "SYSID"
+            , "SYSTEM"
+            , "TABLESPACE"
+            , "TEMP"
+            , "TEMPLATE"
+            , "TEMPORARY"
+            , "TOAST"
+            , "TRANSACTION"
+            , "TRIGGER"
+            , "TRUNCATE"
+            , "TRUSTED"
+            , "TYPE"
+            , "UNCOMMITTED"
+            , "UNENCRYPTED"
+            , "UNKNOWN"
+            , "UNLISTEN"
+            , "UNTIL"
+            , "UPDATE"
+            , "VACUUM"
+            , "VALID"
+            , "VALIDATOR"
+            , "VALUES"
+            , "VARYING"
+            , "VIEW"
+            , "VOLATILE"
+            , "WITH"
+            , "WITHOUT"
+            , "WORK"
+            , "WRITE"
+            , "YEAR"
+            , "ZONE"
+            , "BIGINT"
+            , "BIT"
+            , "BOOLEAN"
+            , "CHAR"
+            , "CHARACTER"
+            , "COALESCE"
+            , "CONVERT"
+            , "DEC"
+            , "DECIMAL"
+            , "EXISTS"
+            , "EXTRACT"
+            , "FLOAT"
+            , "GREATEST"
+            , "INOUT"
+            , "INT"
+            , "INTEGER"
+            , "INTERVAL"
+            , "LEAST"
+            , "NATIONAL"
+            , "NCHAR"
+            , "NONE"
+            , "NULLIF"
+            , "NUMERIC"
+            , "OUT"
+            , "OVERLAY"
+            , "POSITION"
+            , "PRECISION"
+            , "REAL"
+            , "ROW"
+            , "SETOF"
+            , "SMALLINT"
+            , "SUBSTRING"
+            , "TIME"
+            , "TIMESTAMP"
+            , "TREAT"
+            , "TRIM"
+            , "VARCHAR"
+            ]

--- a/IHP/IDE/SchemaDesigner/Controller/EnumValues.hs
+++ b/IHP/IDE/SchemaDesigner/Controller/EnumValues.hs
@@ -35,7 +35,7 @@ instance Controller EnumValuesController where
         let enumName = param "enumName"
         let enum = findEnumByName enumName statements
         let values = maybe [] (get #values) enum
-        let value = removeQuotes (cs (values !! valueId))
+        let value = (cs (values !! valueId))
         render EditEnumValueView { .. }
 
     action UpdateEnumValueAction = do
@@ -61,12 +61,12 @@ instance Controller EnumValuesController where
 
 addValueToEnum :: Text -> Text -> Statement -> Statement
 addValueToEnum enumName enumValueName (table@CreateEnumType { name, values }) | name == enumName =
-    table { values = values <> ["'" <> enumValueName <> "'"] }
+    table { values = values <> [enumValueName] }
 addValueToEnum enumName enumValueName statement = statement
 
 updateValueInEnum :: Text -> Text -> Int -> Statement -> Statement
 updateValueInEnum enumName value valueId (table@CreateEnumType { name, values }) | name == enumName =
-    table { values = (replace valueId ("'" <> value <> "'") values) }
+    table { values = (replace valueId value values) }
 updateValueInEnum enumName value valueId statement = statement
 
 deleteValueInEnum :: Text -> Int -> Statement -> Statement

--- a/IHP/IDE/SchemaDesigner/Controller/Tables.hs
+++ b/IHP/IDE/SchemaDesigner/Controller/Tables.hs
@@ -77,7 +77,7 @@ addTable tableName list = list <> [CreateTable { name = tableName, columns = [Co
                 { name = "id"
                 , columnType = "UUID"
                 , primaryKey = True
-                , defaultValue = "uuid_generate_v4()"
+                , defaultValue = Just (CallExpression "uuid_generate_v4" [])
                 , notNull = True
                 , isUnique = False
                 }] }]

--- a/IHP/IDE/SchemaDesigner/Parser.hs
+++ b/IHP/IDE/SchemaDesigner/Parser.hs
@@ -7,6 +7,7 @@ module IHP.IDE.SchemaDesigner.Parser
 ( parseSchemaSql
 , schemaFilePath
 , parseDDL
+, expression
 ) where
 
 import IHP.Prelude
@@ -78,7 +79,7 @@ createEnumType = do
     name <- identifier
     lexeme "AS"
     lexeme "ENUM"
-    values <- between (char '(' >> space) (char ')' >> space) (stringExpr `sepBy` (char ',' >> space))
+    values <- between (char '(' >> space) (char ')' >> space) (textExpr' `sepBy1` (char ',' >> space))
     char ';'
     pure CreateEnumType { name, values }
 
@@ -119,8 +120,7 @@ column = do
     space
     defaultValue <- optional do
         lexeme "DEFAULT"
-        value <- expression
-        pure (cs value)
+        expression
     primaryKey <- isJust <$> optional (lexeme "PRIMARY" >> lexeme "KEY")
     notNull <- isJust <$> optional (lexeme "NOT" >> lexeme "NULL")
     isUnique <- isJust <$> optional (lexeme "UNIQUE")
@@ -141,23 +141,26 @@ sqlType = choice
         ]
 
 
+expression :: Parser Expression
 expression = do
-    e <- try callExpr <|> varExpr <|> stringExpr 
+    e <- try callExpr <|> varExpr <|> textExpr 
     space 
     pure e
 
-varExpr :: Parser Text
-varExpr = identifier
+varExpr :: Parser Expression
+varExpr = VarExpression <$> identifier
 
+callExpr :: Parser Expression
 callExpr = do
     func <- identifier
     args <- between (char '(') (char ')') (expression `sepBy` char ',')
-    pure (func <> "(" <> intercalate ", " args <> ")")
+    pure (CallExpression func args)
 
-stringExpr :: Parser Text
-stringExpr = do
-    str <- cs <$> (char '\'' *> manyTill Lexer.charLiteral (char '\''))
-    pure ("'" <> str <> "'")
+textExpr :: Parser Expression
+textExpr = TextExpression <$> textExpr'
+
+textExpr' :: Parser Text
+textExpr' = cs <$> (char '\'' *> manyTill Lexer.charLiteral (char '\''))
 
 identifier :: Parser Text
 identifier = do

--- a/IHP/IDE/SchemaDesigner/Types.hs
+++ b/IHP/IDE/SchemaDesigner/Types.hs
@@ -25,7 +25,7 @@ data Column = Column
     { name :: Text
     , columnType :: Text
     , primaryKey :: Bool
-    , defaultValue :: Maybe Text
+    , defaultValue :: Maybe Expression
     , notNull :: Bool
     , isUnique :: Bool
     }
@@ -46,4 +46,13 @@ data Constraint
         , referenceColumn :: Maybe Text
         , onDelete :: Maybe OnDelete
         }
+    deriving (Eq, Show)
+
+data Expression =
+    -- | Sql string like @'hello'@
+    TextExpression Text
+    -- | Simple variable like @users@
+    | VarExpression Text
+    -- | Simple call, like @COALESCE(name, 'unknown name')@
+    | CallExpression Text [Expression]
     deriving (Eq, Show)

--- a/IHP/IDE/SchemaDesigner/View/Columns/Edit.hs
+++ b/IHP/IDE/SchemaDesigner/View/Columns/Edit.hs
@@ -2,6 +2,7 @@ module IHP.IDE.SchemaDesigner.View.Columns.Edit where
 
 import IHP.ViewPrelude
 import IHP.IDE.SchemaDesigner.Types
+import IHP.IDE.SchemaDesigner.Compiler (compileExpression)
 import IHP.IDE.ToolServer.Types
 import IHP.IDE.ToolServer.Layout
 import IHP.View.Modal
@@ -114,29 +115,18 @@ typeSelector selected = preEscapedToHtml [plain|
             then [plain|<option value=#{value} selected>#{text}</option>|]
             else [plain|<option value=#{value}>#{text}</option>|]
 
-defaultSelector :: Maybe Text -> Html
-defaultSelector selected = preEscapedToHtml [plain|
+defaultSelector :: Maybe Expression -> Html
+defaultSelector defValue = preEscapedToHtml [hsx|
     <div class="col-sm-10">
         <select id="defaultSelector" name="defaultValue" class="form-control select2">
-            #{option (selectedType selected) "NODEFAULT" "no default"}
-            #{maybeCustom selected}
+            {forEach values renderValue}
         </select>
     </div>
 |]
     where
-        option selected value text = if selected == value
-            then [plain|<option value=#{tshow (fromMaybe "" value)} selected>#{text}</option>|]
-            else [plain|<option value=#{tshow (fromMaybe "" value)}>#{text}</option>|]
-        selectedType selection = case selection of
-            Just "''" -> "EMPTY"
-            Just "NULL" -> "NULL"
-            Just "null" -> "NULL"
-            Nothing -> "NODEFAULT"
-            custom -> custom
-        maybeCustom selection = case selection of
-            Just "''" -> [plain|<option value="EMPTY" selected>''</option>|]
-            Just "NULL" -> [plain|<option value="NULL" selected>null</option>|]
-            Just "null" -> [plain|<option value="NULL" selected>null</option>|]
-            Just "EMPTY" -> [plain|<option value="EMPTY" selected>''</option>|]
-            Nothing -> mempty
-            _ -> [plain|<option value=#{fromMaybe "" selection} selected>#{fromMaybe "" selection}</option>|]
+        suggestedValues = [Nothing, Just (TextExpression ""), Just (VarExpression "null"), Just (CallExpression "NOW" [])]
+        values = if defValue `elem` suggestedValues then suggestedValues else defValue:suggestedValues
+
+        renderValue :: Maybe Expression -> Html
+        renderValue e@(Just expression) = [hsx|<option value={compileExpression expression} selected={e == defValue}>{compileExpression expression}</option>|]
+        renderValue Nothing = [hsx|<option value="NODEFAULT" selected={Nothing == defValue}>No default</option>|]

--- a/IHP/IDE/SchemaDesigner/View/Layout.hs
+++ b/IHP/IDE/SchemaDesigner/View/Layout.hs
@@ -4,7 +4,9 @@ import IHP.ViewPrelude
 import IHP.IDE.SchemaDesigner.Types
 import IHP.IDE.ToolServer.Types
 import IHP.IDE.ToolServer.Layout
-import IHP.IDE.SchemaDesigner.Compiler (compileIdentifier)
+import IHP.IDE.SchemaDesigner.Compiler (compileIdentifier, compileExpression)
+import qualified IHP.IDE.SchemaDesigner.Parser as Parser
+import qualified Text.Megaparsec as Megaparsec
 import qualified Data.List as List
 
 schemaDesignerLayout :: Html -> Html
@@ -144,7 +146,7 @@ renderColumn Column { name, primaryKey, columnType, defaultValue, notNull, isUni
         renderIsUnique = if isUnique then [hsx|IS UNIQUE|] else mempty
         renderDefault =
             case defaultValue of
-                Just value -> [hsx|default: {value} |]
+                Just value -> [hsx|default: {compileExpression value} |]
                 Nothing -> mempty
         renderForeignKey = case findForeignKey statements tableName name of
             Just addConstraint@AddConstraint { constraint } -> [hsx|<a href={EditForeignKeyAction tableName name (get #constraintName addConstraint) (get #referenceTable constraint)} class="d-block nounderline" style="color: #808080;">FOREIGN KEY: {get #referenceTable constraint}</a>|]
@@ -178,7 +180,7 @@ renderValue :: Text -> Int -> Text -> Html
 renderValue value valueId enumName = [hsx|
 <tr class="column">
     <td class="context-column column-name" oncontextmenu={"showContextMenu('" <> contextMenuId <> "'); event.stopPropagation();"}>
-        {removeQuotes (cs value)}
+        {value}
     </td>
 </tr>
 <div class="custom-menu menu-for-column shadow backdrop-blur" id={contextMenuId}>
@@ -279,26 +281,10 @@ replace :: Int -> a -> [a] -> [a]
 replace i e xs = case List.splitAt i xs of
    (before, _:after) -> before ++ (e: after)
 
-getDefaultValue :: Text -> Text -> Maybe Text
-getDefaultValue columnType value = case value of
-    "EMPTY" -> Just "''"
-    "NULL" -> Just "NULL"
-    "NODEFAULT" -> Nothing
-    "NOW()" -> Just value
-    "uuid_generate_v4()" -> Just value
-    custom -> case columnType of
-        "TEXT" -> Just ("'" <> custom <> "'")
-        "INT" -> Just custom
-        "UUID" -> Just ("'" <> custom <> "'")
-        "BOOLEAN" -> Just custom
-        "TIMESTAMP WITH TIME ZONE" -> Just ("'" <> custom <> "'")
-        "REAL" -> Just custom
-        "DOUBLE PRECISION" -> Just custom
-        "DATE" -> Just ("'" <> custom <> "'")
-        "BINARY" -> Just ("'" <> custom <> "'")
-        "TIME" -> Just ("'" <> custom <> "'")
-        "POINT" -> Just ("'" <> custom <> "'")
-        _ -> Just ("'" <> custom <> "'")
+getDefaultValue :: Text -> Text -> Maybe Expression
+getDefaultValue columnType value = case Megaparsec.runParser Parser.expression "" value of
+        Left _ -> Nothing
+        Right expression -> Just expression
 
 isIllegalKeyword :: Text -> Bool
 isIllegalKeyword input = case (toUpper input) of

--- a/IHP/SchemaCompiler.hs
+++ b/IHP/SchemaCompiler.hs
@@ -317,6 +317,7 @@ compileCreate table@(CreateTable { name, columns }) =
                     )
             )
 
+commaSep :: [Text] -> Text
 commaSep = intercalate ", "
 
 toBinding :: Text -> Column -> Text
@@ -328,7 +329,7 @@ compileUpdate table@(CreateTable { name, columns }) =
         modelName = tableNameToModelName name
         values = commaSep (map toValue columns)
 
-        toValue Column { defaultValue = Just theDefaultValue } = theDefaultValue
+        toValue Column { defaultValue = Just theDefaultValue } = "DEFAULT"
         toValue _ = "?"
 
         bindings :: Text
@@ -347,7 +348,6 @@ compileUpdate table@(CreateTable { name, columns }) =
                     <> "pure (List.head result)\n"
                 )
             )
-
 
 compileFromRowInstance :: (?schema :: Schema) => Statement -> Text
 compileFromRowInstance table@(CreateTable { name, columns }) =
@@ -420,12 +420,16 @@ toDefaultValueExpr Column { columnType, notNull, defaultValue = Just theDefaultV
                 wrapNull False value = "(Just " <> value <> ")"
                 wrapNull True value = value
             in
-                if theDefaultValue == "null"
+                if theDefaultValue == VarExpression "null"
                     then "Nothing"
                     else
                         case columnType of
-                            "TEXT" -> wrapNull notNull (tshow theDefaultValue)
-                            "BOOl" -> wrapNull notNull (tshow (toLower theDefaultValue == "true"))
+                            "TEXT" -> case theDefaultValue of
+                                TextExpression value -> wrapNull notNull (tshow value)
+                                otherwise            -> error ("toDefaultValueExpr: TEXT column needs to have a TextExpression as default value. Got: " <> show otherwise)
+                            "BOOl" -> case theDefaultValue of
+                                VarExpression value -> wrapNull notNull (tshow (toLower value == "true"))
+                                otherwise           -> error ("toDefaultValueExpr: BOOL column needs to have a VarExpression as default value. Got: " <> show otherwise)
                             _ -> "def"
 toDefaultValueExpr _ = "def"
 

--- a/IHP/SchemaCompiler.hs
+++ b/IHP/SchemaCompiler.hs
@@ -260,7 +260,7 @@ compileEnumDataDefinitions enum@(CreateEnumType { name, values }) =
         "data " <> modelName <> " = " <> (intercalate " | " valueConstructors) <> " deriving (Eq, Show, Read, Enum)\n"
         <> "instance FromField " <> modelName <> " where\n"
         <> indent (unlines (map compileFromFieldInstanceForValue values))
-        <> "    fromField field (Just value) = returnError ConversionFailed field \"Unexpected value for enum value\""
+        <> "    fromField field (Just value) = returnError ConversionFailed field \"Unexpected value for enum value\"\n"
         <> "    fromField field Nothing = returnError UnexpectedNull field \"Unexpected null for enum value\"\n"
         <> "instance Default " <> modelName <> " where def = " <> tableNameToModelName (unsafeHead values) <> "\n"
         <> "instance ToField " <> modelName <> " where\n" <> indent (unlines (map compileToFieldInstanceForValue values))

--- a/Test/IDE/SchemaDesigner/CompilerSpec.hs
+++ b/Test/IDE/SchemaDesigner/CompilerSpec.hs
@@ -51,7 +51,7 @@ main = hspec do
                             { name = "id"
                             , columnType = "UUID"
                             , primaryKey = True
-                            , defaultValue = Just "uuid_generate_v4()"
+                            , defaultValue = Just (CallExpression "uuid_generate_v4" [])
                             , notNull = True
                             , isUnique = False
                             }
@@ -107,7 +107,7 @@ main = hspec do
                             { name = "created_at"
                             , columnType = "TIMESTAMP WITH TIME ZONE"
                             , primaryKey = False
-                            , defaultValue = Just "NOW()"
+                            , defaultValue = Just (CallExpression "NOW" [])
                             , notNull = True
                             , isUnique = False
                             }
@@ -116,7 +116,7 @@ main = hspec do
             compileSql [statement] `shouldBe` sql
 
         it "should compile a CREATE TABLE with quoted identifiers" do
-            compileSql [CreateTable { name = tshow "quoted name", columns = [] }] `shouldBe` "CREATE TABLE \"quoted name\" (\n\n);\n"
+            compileSql [CreateTable { name = "quoted name", columns = [] }] `shouldBe` "CREATE TABLE \"quoted name\" (\n\n);\n"
 
         it "should compile ALTER TABLE .. ADD FOREIGN KEY .. ON DELETE CASCADE" do
             let statement = AddConstraint
@@ -182,3 +182,29 @@ main = hspec do
                         }
                     }
             compileSql [statement] `shouldBe` "ALTER TABLE users ADD CONSTRAINT users_ref_company_id FOREIGN KEY (company_id) REFERENCES companies (id) ;\n"
+
+
+        it "should compile a CREATE TABLE with text default value in columns" do
+            let sql = cs [plain|CREATE TABLE a (\n    content TEXT DEFAULT 'example text' NOT NULL\n);\n|]
+            let statement = CreateTable
+                    { name = "a"
+                    , columns = [
+                        Column
+                            { name = "content"
+                            , columnType = "TEXT"
+                            , primaryKey = False
+                            , defaultValue = Just (TextExpression "example text")
+                            , notNull = True
+                            , isUnique = False
+                            }
+                        ]
+                    }
+            compileSql [statement] `shouldBe` sql
+
+        it "should compile a CREATE TYPE .. AS ENUM" do
+            let sql = cs [plain|CREATE TYPE colors AS ENUM ('yellow', 'red', 'blue');\n|]
+            let statement = CreateEnumType
+                    { name = "colors"
+                    , values = ["yellow", "red", "blue"]
+                    }
+            compileSql [statement] `shouldBe` sql           

--- a/Test/IDE/SchemaDesigner/CompilerSpec.hs
+++ b/Test/IDE/SchemaDesigner/CompilerSpec.hs
@@ -1,14 +1,6 @@
 {-|
 Module: Test.IDE.SchemaDesigner.CompilerSpec
 Copyright: (c) digitally induced GmbH, 2020
-
-When in the IHP directory, you can run this file like:
-
- > nix-shell NixSupport/shell.nix
- > ghci
- > :l Test/IDE/SchemaDesigner/CompilerSpec.hs
- > main
-
 -}
 module Test.IDE.SchemaDesigner.CompilerSpec where
 
@@ -20,8 +12,7 @@ import IHP.ViewPrelude (cs, plain)
 import qualified Text.Megaparsec as Megaparsec
 
 
-main :: IO ()
-main = hspec do
+tests = do
     describe "The Schema.sql Compiler" do
         it "should compile an empty CREATE TABLE statement" do
             compileSql [CreateTable { name = "users", columns = [] }] `shouldBe` "CREATE TABLE users (\n\n);\n"

--- a/Test/IDE/SchemaDesigner/ParserSpec.hs
+++ b/Test/IDE/SchemaDesigner/ParserSpec.hs
@@ -1,14 +1,6 @@
 {-|
 Module: Test.IDE.SchemaDesigner.ParserSpec
 Copyright: (c) digitally induced GmbH, 2020
-
-When in the IHP directory, you can run this file like:
-
- > nix-shell NixSupport/shell.nix
- > ghci
- > :l Test/IDE/SchemaDesigner/ParserSpec.hs
- > main
-
 -}
 module Test.IDE.SchemaDesigner.ParserSpec where
 
@@ -21,8 +13,7 @@ import qualified Text.Megaparsec as Megaparsec
 import GHC.IO (evaluate)
 
 
-main :: IO ()
-main = hspec do
+tests = do
     describe "The Schema.sql Parser" do
         it "should parse an empty CREATE TABLE statement" do
             parseSql "CREATE TABLE users ();"  `shouldBe` CreateTable { name = "users", columns = [] }

--- a/Test/IDE/SchemaDesigner/ParserSpec.hs
+++ b/Test/IDE/SchemaDesigner/ParserSpec.hs
@@ -18,6 +18,7 @@ import qualified IHP.IDE.SchemaDesigner.Parser as Parser
 import IHP.IDE.SchemaDesigner.Types
 import IHP.ViewPrelude (cs, plain)
 import qualified Text.Megaparsec as Megaparsec
+import GHC.IO (evaluate)
 
 
 main :: IO ()
@@ -50,7 +51,7 @@ main = hspec do
                             { name = "id"
                             , columnType = "UUID"
                             , primaryKey = True
-                            , defaultValue = Just "uuid_generate_v4()"
+                            , defaultValue = Just (CallExpression "uuid_generate_v4" [])
                             , notNull = True
                             , isUnique = False
                             }
@@ -106,7 +107,7 @@ main = hspec do
                             { name = "created_at"
                             , columnType = "TIMESTAMP WITH TIME ZONE"
                             , primaryKey = False
-                            , defaultValue = Just "NOW()"
+                            , defaultValue = Just (CallExpression "NOW" [])
                             , notNull = True
                             , isUnique = False
                             }
@@ -175,6 +176,12 @@ main = hspec do
                         , onDelete = Nothing
                         }
                     }
+
+        it "should parse CREATE TYPE .. AS ENUM" do
+            parseSql "CREATE TYPE colors AS ENUM ('yellow', 'red', 'green');" `shouldBe` CreateEnumType { name = "colors", values = ["yellow", "red", "green"] }
+        
+        it "should fail on CREATE TYPE .. AS ENUM without values" do
+            evaluate (parseSql "CREATE TYPE colors AS ENUM ();") `shouldThrow` anyErrorCall
 
 parseSql :: Text -> Statement
 parseSql sql = let [statement] = parseSqlStatements sql in statement

--- a/Test/Main.hs
+++ b/Test/Main.hs
@@ -1,0 +1,28 @@
+{-|
+Module: Test.IDE.SchemaDesigner.CompilerSpec
+Description: Entrypoint to the hspec Testsuite
+Copyright: (c) digitally induced GmbH, 2020
+
+When in the IHP directory, you can run this file like:
+
+ > nix-shell NixSupport/shell.nix
+ > ghci
+ > :l Test/Main.hs
+ > main
+
+-}
+module Main where
+
+import Test.Hspec
+import IHP.Prelude
+
+import qualified Test.IDE.SchemaDesigner.CompilerSpec
+import qualified Test.IDE.SchemaDesigner.ParserSpec
+import qualified Test.ValidationSupport.ValidateFieldSpec
+
+
+main :: IO ()
+main = hspec do
+    Test.IDE.SchemaDesigner.CompilerSpec.tests
+    Test.IDE.SchemaDesigner.ParserSpec.tests
+    Test.ValidationSupport.ValidateFieldSpec.tests

--- a/Test/ValidationSupport/ValidateFieldSpec.hs
+++ b/Test/ValidationSupport/ValidateFieldSpec.hs
@@ -1,14 +1,6 @@
 {-|
 Module: Test.ValidationSupport.ValidateFieldSpec
 Copyright: (c) digitally induced GmbH, 2020
-
-When in the IHP directory, you can run this file like:
-
- > nix-shell NixSupport/shell.nix
- > ghci
- > :l Test/ValidationSupport/ValidateFieldSpec.hs
- > main
-
 -}
 module Test.ValidationSupport.ValidateFieldSpec where
 
@@ -19,8 +11,7 @@ import IHP.ValidationSupport.Types
 import IHP.ViewPrelude (cs, plain)
 import Data.Text (Text, length)
 
-main :: IO ()
-main = hspec do
+tests = do
     describe "The nonEmpty validator" do
         it "should handle trivial cases" do
             nonEmpty ("hi" :: Text) `shouldBe` Success

